Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp

```diff
@@ -0,0 +1,149 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebGPUSurfaceCocoa.h"
+
+#include "PlatformCALayer.h"
+#include "pal/graphics/WebGPU/WebGPUCanvasConfiguration.h"
+#include "pal/graphics/WebGPU/WebGPUDevice.h"
+#include "pal/graphics/WebGPU/WebGPUTextureDescriptor.h"
+
+namespace WebCore {
+
+WebGPUSurfaceCocoa::WebGPUSurfaceCocoa(int width, int height, PAL::WebGPU::CanvasConfiguration configuration)
+    : m_width(width)
+    , m_height(height)
+    , m_configuration(WTFMove(configuration))
+{
+}
+
+void WebGPUSurfaceCocoa::prepareToDelegateDisplay(PlatformCALayer&)
+{
+}
+
+void WebGPUSurfaceCocoa::display(PlatformCALayer& layer)
+{
+    if (m_frontBuffer)
+        layer.setContents(m_frontBuffer->surface());
+}
+
+GraphicsLayer::CompositingCoordinatesOrientation WebGPUSurfaceCocoa::orientation() const
+{
+    return GraphicsLayer::CompositingCoordinatesOrientation::TopDown;
+}
+
+void WebGPUSurfaceCocoa::present()
+{
+    std::swap(m_frontBuffer, m_backBuffer);
+    m_backBufferTexture = nullptr;
+    std::swap(m_backBuffer, m_secondaryBackBuffer);
+}
+
+bool WebGPUSurfaceCocoa::isValidBuffer(const IOSurface* buffer)
+{
+    if (!buffer)
+        return false;
+
+    if (buffer->size() != IntSize { m_width, m_height })
+        return false;
+
+    if (buffer->isInUse())
+        return false;
+
+    return true;
+}
+
+void WebGPUSurfaceCocoa::ensureValidBackBuffer()
+{
+    if (isValidBuffer(m_backBuffer.get()))
+        return;
+
+    // If the back buffer is not valid, try to recycle the secondary back buffer.
+    m_backBuffer = WTFMove(m_secondaryBackBuffer);
+    m_backBufferTexture = nullptr;
+    m_secondaryBackBuffer = nullptr;
```

Comment:
You just did a move so this should not be needed.

Reply:
Fixed.